sale_loyalty_order_line_link/models/sale_order.py

```diff
@@ -175,11 +149,15 @@ def _filter_related_program_lines(self, program):
         """Discard those lines not in the program domain. With other modules changing
         the product criteria rules, we can extend this method to return the proper
         records."""
-        # No domain means that any product meets the promotion rules
-        if not program.rule_products_domain:
-            return self
-        domain = safe_eval(program.rule_products_domain)
-        products = self.mapped("product_id").filtered_domain(domain)
-        return self.filtered(
-            lambda x: x.product_id in products and not x.is_reward_line
-        )
+        programs_line = self.env["sale.order.line"]
+        for rule in program.rule_ids:
+            valid_product_domain = rule._get_valid_product_domain()
+            if valid_product_domain:
+                domain = safe_eval(str(valid_product_domain))
+                products = self.mapped("product_id").filtered_domain(domain)
+                programs_line |= self.filtered(
+                    lambda x: x.product_id in products and not x.is_reward_line
+                )
+            else:
+                programs_line |= self.filtered(lambda x: not x.is_reward_line)
+        return programs_line
```

Comment:
The logic was quite simple before and you don't need to make it this complex at all. Just use the previous logic taking the domain from the provided method